comit/src/network/protocols/setup_swap.rs

```diff
@@ -0,0 +1,646 @@
+use crate::{asset, ethereum::ChainId, hbit, herc20, identity, Role, SecretHash, Timestamp};
+use bitcoin::{hashes::core::fmt::Formatter, Network};
+use futures::prelude::*;
+use libp2p::{
+    core::upgrade,
+    request_response::{
+        ProtocolName, ProtocolSupport, RequestResponse, RequestResponseCodec,
+        RequestResponseConfig, RequestResponseEvent, RequestResponseMessage,
+    },
+    swarm::{NetworkBehaviourAction, NetworkBehaviourEventProcess, PollParameters},
+    NetworkBehaviour, PeerId,
+};
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::{HashMap, VecDeque},
+    fmt::{Debug, Display},
+    io,
+    marker::PhantomData,
+    task::{Context, Poll},
+};
+
+/// Event emitted by the `Comit` behaviour.
+#[derive(Clone, Debug)]
+#[allow(clippy::large_enum_variant)]
+pub enum BehaviourOutEvent {
+    Herc20Hbit {
+        alpha: herc20::Params,
+        beta: hbit::Params,
+    },
+    HbitHerc20 {
+        alpha: hbit::Params,
+        beta: herc20::Params,
+    },
+    AlreadyHaveRoleParams {
+        peer: PeerId,
+        have: RoleDependentParams,
+        received: RoleDependentParams,
+    },
+}
+
+#[derive(Clone, Debug, thiserror::Error)]
+#[error("peer {peer} also wants to be {role} for swap {id}")]
+pub struct RoleMistmatch {
+    peer: PeerId,
+    role: Role,
+    id: CommonParams,
+}
+
+impl BehaviourOutEvent {
+    fn new(
+        common: CommonParams,
+        alice: &AliceParams,
+        bob: &BobParams,
+        swap_protocol: SwapProtocol,
+    ) -> BehaviourOutEvent {
+        match swap_protocol {
+            SwapProtocol::HbitHerc20 => BehaviourOutEvent::HbitHerc20 {
+                alpha: hbit::Params {
+                    network: Network::Bitcoin,
+                    asset: common.bitcoin_quantity,
+                    redeem_identity: bob.bitcoin_identity,
+                    refund_identity: alice.bitcoin_identity,
+                    expiry: Timestamp::from(common.bitcoin_absolute_expiry),
+                    secret_hash: alice.secret_hash,
+                },
+                beta: herc20::Params {
+                    asset: asset::Erc20 {
+                        token_contract: Default::default(),
+                        quantity: common.ethereum_quantity,
+                    },
+                    redeem_identity: alice.ethereum_identity,
+                    refund_identity: bob.ethereum_identity,
+                    expiry: Timestamp::from(common.ethereum_absolute_expiry),
+                    secret_hash: alice.secret_hash,
+                    chain_id: ChainId::regtest(),
```

Comment:
I _think_ the chain ID and bitcoin network should be part of the `CommonParams` as well. We can retrieve them from the config when we convert the match and trigger `SetupSwap`. I feel it is worthwhile "asserting" them as well with the other party. What do you think?

Reply:
yep im cool with that, should be at some point and I guess it is related to the swap.